src/common/raft/raft_utils.cuh

```diff
@@ -0,0 +1,130 @@
+    }
+    ~gpu_resources() {
+        memory_resources_.clear();
+    }
+
+    void
+    init(int device_id = get_current_device()) {
+        auto lock = std::lock_guard{raft_mutex};
+        auto stream_iter = stream_pools_.find(device_id);
+
+        if (stream_iter == stream_pools_.end()) {
+            auto scoped_device = device_setter{device_id};
+            stream_pools_[device_id] = std::make_shared<rmm::cuda_stream_pool>(streams_per_device_);
+
+            // Set up device memory pool for this device
+            auto init_pool_size = thrust::optional<std::size_t>{};
+            auto max_pool_size = thrust::optional<std::size_t>{};
+            auto* env_str = getenv("KNOWHERE_GPU_MEM_POOL_SIZE");
+            if (env_str != NULL) {
+                auto init_pool_size_tmp = std::size_t{};
+                auto max_pool_size_tmp = std::size_t{};
+                auto stat = sscanf(env_str, "%zu;%zu", &init_pool_size_tmp, &max_pool_size_tmp);
+                if (stat == 2) {
+                    LOG_KNOWHERE_INFO_ << "Get Gpu Pool Size From env, init size: " << init_pool_size_tmp
+                                       << " MB, max size: " << max_pool_size_tmp << " MB";
+                    init_pool_size = init_pool_size_tmp << 20;
+                    max_pool_size = max_pool_size_tmp << 20;
+                } else {
+                    LOG_KNOWHERE_WARNING_ << "please check env format";
+                }
+            }
+            memory_resources_[device_id] =
+                std::make_unique<rmm::mr::pool_memory_resource<rmm::mr::cuda_memory_resource>>(
+                    &upstream_mr_, init_pool_size, max_pool_size);
+            rmm::mr::set_current_device_resource(memory_resources_[device_id].get());
+        }
+    }
+
+    auto
+    get_streams_per_device() const {
+        return streams_per_device_;
+    }
+
+    auto
+    get_stream_view(int device_id = get_current_device(), std::size_t thread_id = get_thread_id()) {
+        return stream_pools_[device_id]->get_stream(thread_id % streams_per_device_);
+    }
+
+ private:
+    std::size_t streams_per_device_;
+    std::map<int, std::shared_ptr<rmm::cuda_stream_pool>> stream_pools_;
+    std::map<int, std::unique_ptr<rmm::mr::pool_memory_resource<rmm::mr::cuda_memory_resource>>> memory_resources_;
+    rmm::mr::cuda_memory_resource upstream_mr_;
+};
+
+inline auto&
+get_gpu_resources(std::size_t streams_per_device = std::size_t{1}) {
+    static auto resources = gpu_resources{streams_per_device};
+    return resources;
+}
+
+inline void
+init_gpu_resources(std::size_t streams_per_device = std::size_t{1}, int device_id = get_current_device()) {
+    get_gpu_resources(streams_per_device).init(device_id);
+}
+
+inline auto&
+get_raft_resources(int device_id = get_current_device()) {
+    thread_local auto all_resources = std::map<int, std::unique_ptr<raft::device_resources>>{};
```

Comment:
why thread_local ?

Reply:
This allows lazy instantiation of `device_resources` without requiring a lock every time we want to fetch the required `device_resources`. Individual `device_resources` may share the same stream (avoiding the problem this PR is intended to solve), but we can control how the streams and devices are distributed across threads.

Because we need to reference the correct `device_resources` very frequently, blocking all threads on that fetch would be expensive. Lazy instantiation is desirable because we may not know _a priori _ which host threads will need access to which devices.